File: share/share-flink/src/main/java/com/mysql/JdbcReader.java
package com.mysql;

import com.entity.StudentDemo;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

public class JdbcReader extends RichSourceFunction<StudentDemo> {
    private static final Logger logger = LoggerFactory.getLogger(JdbcReader.class);
    private Connection connection = null;
    private PreparedStatement ps = null;
    Config config = ConfigFactory.load("my.properties");

    //该方法主要用于打开数据库连接，下面的ConfigKeys类是获取配置的类
    public void open(Configuration parameters) throws Exception {
        Class.forName(config.getString("jdbc.driver"));//加载数据库驱动
        connection = DriverManager.getConnection(config.getString("jdbc.url"), config.getString("jdbc.user"), config.getString("jdbc.password"));//获取连接
        String sql = "select stuid,stuname,stuaddr,stusex from Student;";
        ps = connection.prepareStatement(sql);
    }

    //执行查询并获取结果
    public void run(SourceContext<StudentDemo> ctx) {
        try {
            ResultSet resultSet = ps.executeQuery();
            while (resultSet.next()) {
                int stuid = resultSet.getInt("stuid");
                String stuname = resultSet.getString("stuname");
                String stuaddr = resultSet.getString("stuaddr");
                String stusex = resultSet.getString("stusex");
                StudentDemo studentDemo = new StudentDemo(stuid, stuname, stuaddr, stusex);
                ctx.collect(studentDemo);//发送结果，结果是tuple2类型，2表示两个元素，可根据实际情况选择
            }
        } catch (Exception e) {
            logger.error("runException:{}", e);
        }

    }

    //关闭数据库连接
    public void cancel() {
        try {
            super.close();
            if (connection != null) {
                connection.close();
            }
            if (ps != null) {
                ps.close();
            }
        } catch (Exception e) {
            logger.error("runException:{}", e);
        }
    }


}

File: lib/lib-flink/lib-flink-devicemonitor/src/main/scala/com/demo/JdbcReader.scala
package com.demo

import java.sql.DriverManager
import java.sql.Connection
import java.sql.DriverManager
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.util
import java.util.Hashtable
import org.apache.flink.configuration.Configuration
import org.apache.flink.streaming.api.functions.source.
{RichSourceFunction, SourceFunction}
import scala.collection.mutable

case class RegionInfo(region: String, value: String, inner_code: String)

class JdbcReader extends RichSourceFunction[RegionInfo] {
  private var connection: Connection = null
  private var ps: PreparedStatement = null
  private var isRunning: Boolean = true

  override def cancel(): Unit = {
    super.close()
    if (connection != null) {
      connection.close()
    }
    if (ps != null) {
      ps.close()
    }
    isRunning = false
  }

  override def run(ctx: SourceFunction.SourceContext[RegionInfo]): Unit = {
    while (isRunning) {
      val resultSet = ps.executeQuery()
      while (resultSet.next()) {
        val rangeInfo = RegionInfo(resultSet.getString("region"), resultSet.getString("value"), resultSet.getString("inner_code"))ctx.collect(rangeInfo)
      }
      //休息2分钟
      Thread.sleep(5000 * 60)
    }
  }

  override def open(parameters: Configuration): Unit = {
    super.open(parameters)
    val driverClass = "com.mysql.jdbc.Driver"
    val dbUrl = "jdbc:mysql://localhost:3306/flink"
    val userName = "root"
    val passWord = "1234"
    connection = DriverManager.getConnection(dbUrl, userName, passWord)
    ps = connection.prepareStatement("select region, value, inner_code fromevent_mapping")
  }
}
